Create mock wireless module test script

Blake-Haydon opened the pull request.

Blake-Haydon committed:
Mock wireless sensor test script

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+# MQTT wireless sensor test script that sends fake randomly generated data
+# To do a single test --> python3 mqtt_wireless_test.py -t 0
```
No need to write this! It is assumed to the user that they will need to type in `python3 mock_wireless_module --help` that they will be shown how to use this script!

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+import json
+import paho.mqtt.client as mqtt
+
+class Sensors:
```
Good start on writing a class! However, I would move this out to it's own file where the base class is called `MockSensor` and then you have the other Sensors (CO2, humidity etc.) inherit from this base class.

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+
+    def CO2(val=325):
+        # value in ppm (normal background conentration 250-400ppm)
+        return round(val + val*random.uniform(-Sensors.percentRange, Sensors.percentRange), Sensors.decimals)
```
Might want to look into running your code through `pylint` as it will help you in reducing code width!

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+
+    def humidity(val=75):
+        # value as a percentage of humidity
+        return round(val + val*random.uniform(-Sensors.percentRange, Sensors.percentRange), Sensors.decimals)
```
Would be good to split this up!

``` python
def humidity(val=75):
        random_humidity = val + val*random.uniform(-Sensors.percentRange, Sensors.percentRange)
        return round(random_humidity, Sensors.decimals)
```
Like all humans, we're all lazy. If I were to skim this code, I would be able to read what is returned from this function. It also means I don't have t write unnecessary comments to the code too as the code itself its verbose already.

emilytrau replied:
This line `round(val + val*random.uniform(-Sensors.percentRange, Sensors.percentRange), Sensors.decimals)` is repeated a lot. It might be a good idea to refactor this into its own function.

Blake-Haydon replied:
Yeah I found that as well, now it is just one function that gets called every time

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+class Sensors:
+
+    decimals = 2            # number of decimal places for the random generated val
+    percentRange = 0.05     # percent range for the rangomly generated value
```
Python uses snake case for variable names. `percent_range`

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+            JSON_data = json.dumps(data)
+            pub_path = "v3/wireless-sensor/3/data"
+
+            client.publish(pub_path, JSON_data)
```
I think you can bring these two lines out from the if statement as they're just repeated!

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+        if sensorID == 1 or sensorID == None:
+
+            # generate JSON data and publish it over MQTT
+            data = {
```
I like how easy it is to read the output of the data from the wireless sensors!

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+            break
+
+def start_publishing(client, args):
+    print("publishing started...")
```
These MQTT functions should be in it's own class. But to keep this PR short, we won't worry about it for now!

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+def on_message(client, userdata, msg):
+    print(msg.topic+" "+str(msg.payload.decode("utf-8")))
+
+def run():
```
This should be `if __name__ == "__main__":` instead. With python, if you run this script directly, it will run whatever is in the if statement.

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+    start_publishing(client, args)
+    client.loop_stop()
+
+run()
```
This causes the `run()` function to execute if you import the python script in another python script.

khanguslee reviewed (changes requested):
Good work on your first PR!!! There's quite a few changes to make since the PR is pretty long since I got you after you were done. Usually, we like to keep the PR smaller so you don't need to make as much changes and it takes less time for people to review too! Let me know if you have any questions!

I would read https://github.com/monash-human-power/BOOST/blob/master/mqtt_max_speed.py to help you with bunching up the `argparse` stuff

khanguslee commented:
Also, feel free to add the whole of software team to be reviewers of this PR!

khanguslee commented:
When you submit a PR, your PR should also include a description of what the changes were, how to test it and expected results.

Blake-Haydon committed:
Removed header documentation

Blake-Haydon committed:
Changed "==" to "is" for if statements

Blake-Haydon committed:
Remove empty lines

Blake-Haydon committed:
Remove run function

Blake-Haydon committed:
Move Sensor class to MockSensor.py

hallgchris commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+        return {
+        "x": round(val + val*random.uniform(-Sensors.percentRange, Sensors.percentRange), Sensors.decimals),
+        "y": round(val + val*random.uniform(-Sensors.percentRange, Sensors.percentRange), Sensors.decimals),
+        "z": round(val + val*random.uniform(-Sensors.percentRange, Sensors.percentRange), Sensors.decimals)}
```
Just for readability, I'd format this (and future occurrences) as
```python
return {
    "x": round(...),
    "y": round(...),
    "z": round(...),
}
```
Note the indentation, brace on newline, and (optional but I like it) comma on the `"z"` line

hallgchris commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+    percent_range = 0.05
+
+    def CO2(val=325):
+        # value in ppm (normal background conentration 250-400ppm)
```
When adding documentation for a function, class or file, we add comments in a special way so that editors can recognise that this comment documents exactly what this function does, and show it as a tooltip when you hover over the function or whatever. Do this by using triple quote strings, i.e.
```python
def CO2(val=325):
    """ Value in ppm (normal background conentration 250-400ppm) """
```
This applies to all the other function documentation here. Also, typo in concentration

hallgchris commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,170 @@
+
+    # number of decimal places for the random generated val
+    decimals = 2
+    # percent range for the rangomly generated value
```
Typo in randomly

Blake-Haydon committed:
Refactor parser for readability

hallgchris reviewed (changes requested):
Will continue review on new files (GitHub got upset cos the code was all moved around)

hallgchris commented on Raspi/Test/MockSensor.py:
```diff
@@ -0,0 +1,55 @@
+        # value as a percentage of the battery
+        return round(val + val*random.uniform(-Sensors.percent_range, Sensors.percent_range), Sensors.decimals)
+
+    def reedSwitch(val=True):
```
Use `snake_case`

hallgchris commented on Raspi/Test/MockSensor.py:
```diff
@@ -0,0 +1,55 @@
+        # value as either true or false
+        return val
+
+    def GPS(latLngVal=20, speedVal=50):
```
`snake_case`

hallgchris commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,135 @@
+                "accelerometer": Sensors.accelerometer(),
+                "gyroscope": Sensors.gyroscope(),
+                "battery": Sensors.battery(),
+                "time": Sensors.time()}
```
Probably a personal thing but I reckon chuck the brace on a new line

hallgchris replied:
Perhaps I missed something, but aren't we keeping track of time on the DAS pi rather than from the sensors? @khanguslee

khanguslee replied:
Yes correct! Remove time from all modules.

hallgchris commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,135 @@
+    active.""")
+
+
+def send_fake_data(client, duration, rate, sensorID):
```
`snake_case`

hallgchris commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,135 @@
+                "battery": Sensors.battery(),
+                "time": Sensors.time()}
+
+            JSON_data = json.dumps(data)
```
`snake_case` (sorry this is probably annoying lol)

Blake-Haydon replied:
It's good it makes it more readable, thanks for pointing it out

Blake-Haydon committed:
Add init to MockSensor Class

Blake-Haydon committed:
Add get_value method

Blake-Haydon committed:
Add random library

Blake-Haydon committed:
Split random value generation to other func

Blake-Haydon committed:
Remove random library

Blake-Haydon committed:
Refactor parser by removing newlines

Blake-Haydon committed:
Fix indexing error with *ave_val

Blake-Haydon committed:
Import MockSensor class

Blake-Haydon committed:
Add new "Mock sensors"

Blake-Haydon committed:
Fix incorrect snake_case

Blake-Haydon committed:
Refactor test code with the new MockSensor Class

emilytrau commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,160 @@
+            publish(client, battery_topic, battery_data)
+
+        time.sleep(1/rate)
+        if total_time >= duration:
```
This is the same as `while total_time < duration:`

Blake-Haydon replied:
good point, I just used the code from the DAS MQTT test so that statement was in there

Blake-Haydon committed:
Fix incorrect sensor ID for sensor 3

Blake-Haydon committed:
Swap type check for "isinstance()"

Blake-Haydon committed:
Convert confusing if statement to easier while loop

hallgchris commented on Raspi/Test/MockSensor.py:
```diff
@@ -1,55 +1,58 @@
+                                 float""")
+            self.single_val = True
+
+        if len(ave_val) != 1:
```
Maybe use `else` here so that it's clear that exactly one of these if statements will always execute

hallgchris commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -1,112 +1,136 @@
+                   ("course", 0))
+
+
+def send_fake_data(client, duration, rate, sensor_ID):
```
ID should be lowercase (`sensor_id`)

(sorry :stuck_out_tongue_winking_eye:)

hallgchris commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -1,112 +1,136 @@
+    data:   python dict containin the data to be published on the topic
+    """
+    # generate JSON from the python dict
+    JSON_data = json.dumps(data)
```
`json_data`

hallgchris commented on Raspi/Test/MockSensor.py:
```diff
@@ -1,55 +1,58 @@
+        if len(ave_val) == 1:
+            if (not isinstance(ave_val[0], int)
+               and not isinstance(ave_val[0], float)):
+                raise ValueError("""Single values must be either and int or a
```
Should raise a `TypeError`. See https://stackoverflow.com/questions/48343387/valueerror-and-typeerror-in-python

Also change next `ValueError` below

hallgchris reviewed (changes requested):
This is looking really nice! Great job Blake :)

Just a few more small things

khanguslee commented on Raspi/Test/MockSensor.py:
```diff
@@ -0,0 +1,58 @@
+    """ Base class to make a mock sensor that produces random data"""
+
+    def __init__(self, *ave_val, percent_range=0.05, decimals=2):
+        """
```
Liking the boost in documentation this time around!

khanguslee commented on Raspi/Test/MockSensor.py:
```diff
@@ -0,0 +1,58 @@
+    def __init__(self, *ave_val, percent_range=0.05, decimals=2):
+        """
+        ave_val:        Either a single val that sets the average value for the
+                        sensor or and array of subvals formatted as
```
Is this meant to be: "sensor or **an** array of subvals formatted as"

Blake-Haydon replied:
Yes it is meant to be

khanguslee commented on Raspi/Test/MockSensor.py:
```diff
@@ -0,0 +1,58 @@
+        ave_val:        Either a single val that sets the average value for the
+                        sensor or and array of subvals formatted as
+                        (sub_value_name, ave_val). Where sub_value_name is the
+                        name that will be used in the outputed dictionary and
```
spelling: outputted

khanguslee commented on Raspi/Test/MockSensor.py:
```diff
@@ -0,0 +1,58 @@
+class MockSensor:
+    """ Base class to make a mock sensor that produces random data"""
+
+    def __init__(self, *ave_val, percent_range=0.05, decimals=2):
```
I think I would prefer to rename the variable `ave_val` to `average_value`. Just because, at a glance, I did not know what `ave_val` meant. I had to scroll up to this documentation to understand what it represented too. You spend more time reading code than you do writing code, so make it easy to read!

Blake-Haydon replied:
I agree, it would be confusing if you don't know what it stands for.

khanguslee commented on Raspi/Test/MockSensor.py:
```diff
@@ -0,0 +1,58 @@
+            self.single_val = True
+
+        if len(ave_val) != 1:
+            for sub_val in ave_val:
```
Expand the `sub_val` name.

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,159 @@
+import json
+import paho.mqtt.client as mqtt
+
+# my library
```
No need for this comment!

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,159 @@
+
+        # Wireless Sensor 1 (Middle)
+        if sensor_ID == 1 or sensor_ID is None:
+            sensor_data = {
```
The message structure has been changed. I am happy to ignore that for now and for you to fix that next time around! (Just add a `# TODO: Update sensor data with proper structure` to the right lines

khanguslee replied:
Another TODO would be to create another function that actually outputs the wireless module data output. This will make it a lot easier to write tests! Also happy for you to fix this in another PR!

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,159 @@
+    Publishes python dict data to a specific topic in JSON and prints it out
+    client: MQTT client object
+    topic:  MQTT topic eg. '/v3/wireless-module/<id>/start'
+    data:   python dict containin the data to be published on the topic
```
spelling: containing.

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,159 @@
+    topic:  MQTT topic eg. '/v3/wireless-module/<id>/start'
+    data:   python dict containin the data to be published on the topic
+    """
+    # generate JSON from the python dict
```
Comments should start with Capital letter. See below for another instance.

khanguslee reviewed (changes requested):
Good work on the changes! Narrowing down on the problems! Some debt that has to be made after this PR is pulled in though!

khanguslee commented:
Oh, also. Please rename the title of this Pull Request. Example: Create mock wireless module test script. Note that this is what will be shown on the git history when it is merged in!

Blake-Haydon committed:
Remove 'my library' comment

Blake-Haydon committed:
add TODO for outputting random sensor data

Blake-Haydon committed:
Change if state to else for clarity

Blake-Haydon committed:
Replace 'sensor_ID' with 'sensor_id'

Blake-Haydon committed:
Replace 'JSON_data' with 'json_data'

Blake-Haydon committed:
Replace 'ValueError' with 'TypeError'

Blake-Haydon committed:
Fix documentation by changing 'and' to 'an'

'Either a single val that sets the average value for the sensor or and array of subvals' was changed to 'Either a single val that sets the average value for the sensor or an array of subvals'

Blake-Haydon committed:
Fix spelling of 'outputted'

Blake-Haydon committed:
Change 'ave_val' to 'average_value'

Blake-Haydon committed:
Change 'sub_val' to 'sub_value'

Blake-Haydon committed:
Add todo to update message structure

Blake-Haydon committed:
Fix spelling of 'containing'

Blake-Haydon committed:
Capitalise comments

emilytrau reviewed (dismissed):
Looks good 👍

hallgchris reviewed (dismissed):
Looking great! Sorry for the long review haha

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,163 @@
+    help="""Rate of data in data sent per second. Default is 1 data pack sent
+    per second.""")
+parser.add_argument(
+    '--host', action='store', type=str, default="192.168.100.100",
```
Set this to `localhost`.  Don't forget to change the comment below too!

khanguslee commented on Raspi/Test/mock_wireless_module.py:
```diff
@@ -0,0 +1,163 @@
+    help="""Address of the MQTT broker. If nothing is selected it will
+    default to 192.168.100.100.""")
+parser.add_argument(
+    '--id', action='store', type=int, default=None,
```
Include a `-i` alternative as well.

khanguslee reviewed (changes requested):
Some small stuff that I didn't pick up before oops.

Blake-Haydon committed:
Add -i alternative for --id

Blake-Haydon committed:
Change host from '192.168.100.100' to 'localhost'

Blake-Haydon committed:
Fix proper data structure and correct variable naming

c02 was changed to co2 which is what it is meant to be. GPS was also changed to lowercase to match the formating guide

khanguslee reviewed (approved):
Good work! Thanks for all the changes! Hopefully now you know a little bit more about PRs and Python! 👍
